records/generator.go

```diff
@@ -232,24 +258,41 @@ func (rg *RecordGenerator) InsertState(sj StateJSON, domain string, ns string,
 				continue
 			}
 
-			tname := labels.AsDNS952(task.Name)
-			sid := slaveIdTail(task.SlaveId)
-			tag := hashString(task.Id)
-			tail := fname + "." + domain + "."
+			// context used to resolve DomainPattern variables for the current task
+			context := PatternContext{
+				SlaveId: slaveIdTail(task.SlaveId),
+				TaskId: hashString(task.Id),
+				TaskName: labels.AsDNS952(task.Name),
+			}
 
-			// A records for task and task-sid
-			arec := tname + "." + tail
-			rg.insertRR(arec, host, "A")
-			trec := tname + "-" + tag + "-" + sid + "." + tail
+			// task-id A record
+			trec := context.TaskName + "-" + context.TaskId + "-" + context.SlaveId + "." + tail
 			rg.insertRR(trec, host, "A")
 
+			// A records
+			for _, tmpl := range domainTemplates {
+				// skip entries which had an invalid pattern
+				if tmpl == nil {
+					continue
+				}
+
+				// apply pattern to the current record
+				var buf bytes.Buffer
+				err := tmpl.Execute(&buf, context)
+				if err != nil {
+					logging.VeryVerbose.Println("error applying template")
+					continue
+				}
+				rg.insertRR(buf.String()+"."+tail, host, "A")
+			}
+
 			// SRV records
-			if task.Resources.Ports != "" {
+			if task.Resources.Ports != "" && trec != "" {
 				ports := yankPorts(task.Resources.Ports)
 				for _, port := range ports {
 					var srvhost string = trec + ":" + port
-					tcp := "_" + tname + "._tcp." + tail
-					udp := "_" + tname + "._udp." + tail
+					tcp := "_" + context.TaskName + "._tcp." + tail
```

Comment:
i'd actually like for us to shift our thinking re: using dynamic _service types. using standardized service names (even if they're our standard) can provide us some flexibility later on w/ respect to discovery. for example instead of the above:
```
_{taskName}._{proto}.{tail}
```
what about this:
```
_ep._{proto}.{taskName}.{tail}
```
and once we have service discovery, we can get create additional, named endpoint records:
```
_named._{proto}.{portName}.{taskName}.{tail}
--or--
_named._sub._ep._{proto}.{portName}.{taskName}.{tail}
```
(the `_ep` represents a task "endpoint")

or perhaps even this should be configurable via templates. thoughts?